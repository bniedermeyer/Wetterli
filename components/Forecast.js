import React, { Component } from 'react';
import {
  View,
  Text,
  StyleSheet,
  ActivityIndicator,
  Linking
} from 'react-native';

export default class Forecast extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      forecast: undefined,
      location: undefined
    };
  }

  componentDidMount() {
    this.determineGeolocation();
  }

  determineGeolocation() {
    navigator.geolocation.getCurrentPosition(location =>
      this.setState({
        loading: false,
        location: {
          latitude: location.coords.latitude,
          longitude: location.coords.longitude
        }
      })
    );
  }

  async getForecast() {
    if (this.state.location) {
      console.log('requesting forecast');
      const reqUrl = `https://shiny-rc2.begin.app/forecast?lat=${this.state.location.latitude}&lng=${this.state.location.longitude}`;
      try {
        let response = await fetch(reqUrl);
        let responseJson = await response.json();
        this.setState({ forecast: responseJson });
      } catch (error) {
        console.error(error);
      }
    }
  }

  render() {
    if (this.state.loading) {
      return (
        <View>
          <ActivityIndicator size="large" color="#ff8906" />
        </View>
      );
    } else {
      if (!this.state.forecast) {
        this.getForecast();
      }

      return (
        <View
          style={{ flex: 1, justifyContent: 'center', alignContent: 'center' }}
        >
          <Text style={styles.forcastText}>
            {this.state.forecast ? this.state.forecast.temperature : ''}
          </Text>

          <Text style={styles.forcastText}>
            {this.state.forecast ? this.state.forecast.summary : ''}
          </Text>
          <Text
            style={styles.darkSkyLink}
            onPress={() => Linking.openURL('https://darksky.net/poweredby')}
          >
            Powered by Dark Sky
          </Text>
        </View>
      );
    }
  }
}

const styles = StyleSheet.create({
  forcastText: {
    color: '#fff',
    fontSize: 30,
    textAlign: 'center'
  },
  darkSkyLink: {
    color: '#f25f4c'
  }
});
