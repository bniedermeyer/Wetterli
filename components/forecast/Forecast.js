import React, { Component } from 'react';
import { View, ActivityIndicator, Button } from 'react-native';
import ForecastText from './ForecastText';
import ForecastIcon from './ForecastIcon';

export default class Forecast extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      forecast: undefined,
      location: undefined
    };
  }

  componentDidMount() {
    this.determineGeolocation();
  }

  /**
   * Gets the users current location and updates the state
   */
  determineGeolocation() {
    navigator.geolocation.getCurrentPosition(location =>
      this.setState({
        loading: false,
        location: {
          latitude: location.coords.latitude,
          longitude: location.coords.longitude
        }
      })
    );
  }

  /**
   * Using the users current location, make a request to the backend for the current weather report
   */
  async getForecast() {
    if (this.state.location) {
      console.log('requesting forecast');
      const reqUrl = `https://shiny-rc2.begin.app/forecast?lat=${this.state.location.latitude}&lng=${this.state.location.longitude}`;
      try {
        let response = await fetch(reqUrl);
        let responseJson = await response.json();
        this.setState({ forecast: responseJson });
      } catch (error) {
        console.error(error);
      }
    }
  }

  render() {
    if (this.state.loading) {
      return (
        <View>
          <ActivityIndicator size="large" color="#ff8906" />
        </View>
      );
    } else {
      if (!this.state.forecast) {
        this.getForecast(); // only get the forecast if it doesn't currently exist
      }

      return (
        <View
          style={{
            flex: 1,
            justifyContent: 'center',
            alignContent: 'flex-end'
          }}
        >
          <ForecastIcon
            icon={this.state.forecast ? this.state.forecast.icon : undefined}
          />
          <ForecastText forecast={this.state.forecast} />
        </View>
      );
    }
  }
}
