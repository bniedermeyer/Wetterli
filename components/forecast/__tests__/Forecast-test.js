import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Forecast from '../Forecast';

describe('ForecastComponent', () => {
  beforeEach(() => {
    const mockGeolocation = {
      watchPosition: jest.fn(),
      getCurrentPosition: jest.fn(() => ({
        coords: { longitude: 0, latitude: 0 }
      }))
    };

    navigator.geolocation = mockGeolocation;
  });

  it(`renders correctly as loading`, () => {
    const tree = renderer.create(<Forecast />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when making requests for weather', () => {
    const mockResponse = {
      apparentTemperature: 38.14,
      cloudCover: 0.69,
      dewPoint: 37.12,
      humidity: 0.88,
      icon: 'partly-cloudy-day',
      nearestStormBearing: 178,
      nearestStormDistance: 5,
      ozone: 385.9,
      precipIntensity: 0,
      precipProbability: 0,
      pressure: 1009.6,
      summary: 'Mostly Cloudy',
      temperature: 40.47,
      time: 1577038353,
      uvIndex: 0,
      visibility: 10,
      windBearing: 352,
      windGust: 5.98,
      windSpeed: 3.61
    };
    const wrapper = shallow(<Forecast />);
    wrapper.setState({ loading: false, forecast: mockResponse });
    expect(wrapper).toMatchSnapshot();
  });
});
