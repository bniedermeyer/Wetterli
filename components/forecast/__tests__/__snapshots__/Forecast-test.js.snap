// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastComponent renders correctly as loading 1`] = `
<View>
  <ActivityIndicator
    animating={true}
    color="#ff8906"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`ForecastComponent renders correctly when making requests for weather 1`] = `
<View
  style={
    Object {
      "alignContent": "flex-end",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ForecastIcon
    icon="partly-cloudy-day"
  />
  <ForecastText
    forecast={
      Object {
        "apparentTemperature": 38.14,
        "cloudCover": 0.69,
        "dewPoint": 37.12,
        "humidity": 0.88,
        "icon": "partly-cloudy-day",
        "nearestStormBearing": 178,
        "nearestStormDistance": 5,
        "ozone": 385.9,
        "precipIntensity": 0,
        "precipProbability": 0,
        "pressure": 1009.6,
        "summary": "Mostly Cloudy",
        "temperature": 40.47,
        "time": 1577038353,
        "uvIndex": 0,
        "visibility": 10,
        "windBearing": 352,
        "windGust": 5.98,
        "windSpeed": 3.61,
      }
    }
  />
</View>
`;
